### PR TITLE
ci: adds a github action to inspect source code for security problems

### LINKED ISSUE #1305: consider running gosec as a CI Github action to inspect source code for security problems
The `gosec` tool details can be found at https://github.com/securego/gosec.
Suggestion would be to use "continue-on-error" even if there are issues found in the short term

### PR DESCRIPTION
**Pull request by vjayaramrh**
This PR closes #1305 and adds an action to run gosec to inspect source code.
Initially, would like CI to still pass even if issues are found.

Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
Looks like something is not working right:
Run pushd go-controller
~/work/ovn-kubernetes/ovn-kubernetes/go-controller ~/work/ovn-kubernetes/ovn-kubernetes
./hack/gosec.sh: line 2: /home/runner/go/bin/gosec: No such file or directory
##[error]Makefile:49: recipe for target 'gosec' failed
make: *** [gosec] Error 127

**Comment by vjayaramrh** [comment]
> Looks like something is not working right:
> Run pushd go-controller
> ~/work/ovn-kubernetes/ovn-kubernetes/go-controller ~/work/ovn-kubernetes/ovn-kubernetes
> ./hack/gosec.sh: line 2: /home/runner/go/bin/gosec: No such file or directory
> ##[error]Makefile:49: recipe for target 'gosec' failed
> make: *** [gosec] Error 127

Thanks @trozet , shall look into this

**Comment by vjayaramrh** [comment]
> Looks like something is not working right:
> Run pushd go-controller
> ~/work/ovn-kubernetes/ovn-kubernetes/go-controller ~/work/ovn-kubernetes/ovn-kubernetes
> ./hack/gosec.sh: line 2: /home/runner/go/bin/gosec: No such file or directory
> ##[error]Makefile:49: recipe for target 'gosec' failed
> make: *** [gosec] Error 127

@trozet Figured out the issue, there were spaces in Makefile instead of tab

**Comment by danwinship** [comment]
So does this currently generate any warnings? My past experience with `gosec` is that it tends to generate useless and/or incorrect recommendations for code that isn't a web server.

**Comment by vjayaramrh** [comment]
> So does this currently generate any warnings? My past experience with `gosec` is that it tends to generate useless and/or incorrect recommendations for code that isn't a web server.

Yes, currently it does list issues (about 14 which is not too bad), go to link https://github.com/ovn-org/ovn-kubernetes/pull/1308/checks?check_run_id=623710682 and expand `Source code inspection` to look up the details. Also, we as a team can determine which of the available rules (look up Available rules at https://github.com/securego/gosec) are relevant that we want to enable. In the current run, all rules are enabled by default.

**Comment by vjayaramrh** [comment]
recheck

**Comment by danwinship** [comment]
```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:846] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(filepath.Join(saPath, kubeServiceAccountFileToken))
```

This is wrong and really only applies to the case of writing a web server. (Discussed in the earlier cluster-network-operator bug here: https://github.com/openshift/cluster-network-operator/pull/228#discussion_r301641344)

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/nicstobridge.go:133] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(defaultFile)
```

likewise

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/nicstobridge.go:150] - G302 (CWE-276): Expect file permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
  > os.OpenFile(defaultFile, os.O_APPEND|os.O_WRONLY, 0644)
```

This is making assumptions about the nature of the file that we're writing which again are not correct for this use case. (Other files in `/etc/default` have mode `0644` so it makes sense that this file should as well. Well, I mean, ignoring the fact that this whole function is insane.)

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/ovs.go:351] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(pidfile)
```

wrong

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovnkube/ovnkube.go:154] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(pidfile)
```

wrong

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/node.go:107] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(runDir + "ovn-controller.pid")
```

wrong

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:1140] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > os.Open(configFile)
```

wrong

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/cni/helper_linux.go:316] - G204 (CWE-78): Subprocess launched with variable (Confidence: HIGH, Severity: MEDIUM)
  > exec.Command("ovs-vsctl", ovsArgs...)
```

wrong

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:333] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > os.Unsetenv("K8S_APISERVER")


[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:334] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > os.Unsetenv("K8S_TOKEN")

[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:332] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > os.Unsetenv("K8S_CACERT")


[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:331] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > os.Unsetenv("KUBECONFIG")
```

Technically correct according to code inspection (although `os.Unsetenv` never actually returns an error on linux). Not sure why golangci-lint didn't pick this up?

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/cni/cniserver_linux.go:76] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > l.Close()
```

Correct, though irrelevant. (It's a cleanup in an error case, there's nothing interesting to be done if `Close` fails.) Again, not clear why golangci-lint didn't catch this...

```
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go:37] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
  > e.Print()
```

Correct, though there's nothing useful that could be done on error.


If you assume that ovn-kubernetes is a web server, then gosec caught some important bugs. But since we're not a web server, it was wrong about almost everything, and the few things it was right about are things I thought our linter was supposed to be catching already...

**Comment by vjayaramrh** [comment]
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:846] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > ioutil.ReadFile(filepath.Join(saPath, kubeServiceAccountFileToken))
> ```
> 
> 
> This is wrong and really only applies to the case of writing a web server. (Discussed in the earlier cluster-network-operator bug here: [openshift/cluster-network-operator#228 (comment)](https://github.com/openshift/cluster-network-operator/pull/228#discussion_r301641344))
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/nicstobridge.go:133] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > ioutil.ReadFile(defaultFile)
> ```
> 
> 
> likewise
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/nicstobridge.go:150] - G302 (CWE-276): Expect file permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
>   > os.OpenFile(defaultFile, os.O_APPEND|os.O_WRONLY, 0644)
> ```
> 
> 
> This is making assumptions about the nature of the file that we're writing which again are not correct for this use case. (Other files in `/etc/default` have mode `0644` so it makes sense that this file should as well. Well, I mean, ignoring the fact that this whole function is insane.)
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/ovs.go:351] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > ioutil.ReadFile(pidfile)
> ```
> 
> 
> wrong
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovnkube/ovnkube.go:154] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > ioutil.ReadFile(pidfile)
> ```
> 
> 
> wrong
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/node.go:107] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > ioutil.ReadFile(runDir + "ovn-controller.pid")
> ```
> 
> 
> wrong
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:1140] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
>   > os.Open(configFile)
> ```
> 
> 
> wrong
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/cni/helper_linux.go:316] - G204 (CWE-78): Subprocess launched with variable (Confidence: HIGH, Severity: MEDIUM)
>   > exec.Command("ovs-vsctl", ovsArgs...)
> ```
> 
> 
> wrong
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:333] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > os.Unsetenv("K8S_APISERVER")
> 
> 
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:334] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > os.Unsetenv("K8S_TOKEN")
> 
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:332] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > os.Unsetenv("K8S_CACERT")
> 
> 
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config/config.go:331] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > os.Unsetenv("KUBECONFIG")
> ```
> 
> 
> Technically correct according to code inspection (although `os.Unsetenv` never actually returns an error on linux). Not sure why golangci-lint didn't pick this up?
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/cni/cniserver_linux.go:76] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > l.Close()
> ```
> 
> 
> Correct, though irrelevant. (It's a cleanup in an error case, there's nothing interesting to be done if `Close` fails.) Again, not clear why golangci-lint didn't catch this...
> 
> 
> ```
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go:37] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
>   > e.Print()
> ```
> 
> 
> Correct, though there's nothing useful that could be done on error.
> 
> 
> If you assume that ovn-kubernetes is a web server, then gosec caught some important bugs. But since we're not a web server, it was wrong about almost everything, and the few things it was right about are things I thought our linter was supposed to be catching already...

@danwinship in my previous go project, the team decided to move to using secure/gosec as it was configurable in terms of what rules they want included/excluded. The gosec linter as part of golangci-lint did not seem to be catching issues per the teams observation. One of the team members went through the exact exercise that you went through at regular intervals, screening the issues raised by gosec and raising tickets for valid issues that needed to be fixed. The securego/gosec screening was enabled early, so that closer to a release time, team did not have to deal with lot of issues. Per your screening of the fourteen issues raised, it seems like that we do not need to address majority of them which is good news

File: go-controller/hack/gosec.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
+GO111MODULE=on ${GOPATH}/bin/gosec ./...
```
Comment (nerdalert) [inline go-controller/hack/gosec.sh:2]: Might be worth adding a comment on where to get gosec for anyone wanting to run it outside of CI. 

The quickest install I saw was:

```
curl -sfL https://raw.githubusercontent.com/securego/gosec/master/install.sh | sh -s -- -b $GOPATH/bin latest
```

**Comment by nerdalert** [comment]
Gave it a spin. It only took ~10 seconds to run and looks like some worthwhile info. Is it possible to blacklist false positives that Dan mentioned? LGTM other than maybe adding a bit of information on installing gosec.

**Comment by trozet** [comment]
@danwinship so is your opinion here that gosec doesn't give us anything? I've never used it, but I see some of these *could* be useful:
https://github.com/securego/gosec#available-rules

G402: Look for bad TLS connection settings
G404: Insecure random number source (rand)
G303: Creating tempfile using a predictable path
G101: Look for hard coded credentials

Not sure if it will even detect TLS connection settings for ovsdb or anything like that, so might be useless.

**Comment by danwinship** [comment]
I mean, the ignored-error warnings are at least potentially useful, but like I said, I thought our linter was supposed to be catching those already.

> I see some of these _could_ be useful:

but none of those turned up in Vish's run, so either they're not actually useful (eg, we aren't using low-level enough TLS routines that it can tell if we're using them right) or we're smart enough to not make those mistakes... :sunglasses: 

>  G303: Creating tempfile using a predictable path

(This is mostly not a problem anyway since, being in a container, we have our own private copy of /tmp.)

**Comment by trozet** [comment]
@danwinship according to
https://securego.io/docs/rules/g304.html

which flagged:
[/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovnkube/ovnkube.go:154] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > ioutil.ReadFile(pidfile)

we should have been doing:
ioutil.ReadFile(filepath.Clean(pidfile))

I guess pidfile is not really at security risk. But if someone provided a different config file or something like that then maybe? Not really sure if it even applies since we are in a container anyway, but is it worth the minimal effort just to be safe? I'm kind of on the fence, but we should resolve this PR either way.

**Comment by vjayaramrh** [comment]
> @danwinship according to
> https://securego.io/docs/rules/g304.html
> 
> which flagged:
> [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/cmd/ovnkube/ovnkube.go:154] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
> 
> > ioutil.ReadFile(pidfile)
> 
> we should have been doing:
> ioutil.ReadFile(filepath.Clean(pidfile))
> 
> I guess pidfile is not really at security risk. But if someone provided a different config file or something like that then maybe? Not really sure if it even applies since we are in a container anyway, but is it worth the minimal effort just to be safe? I'm kind of on the fence, but we should resolve this PR either way.

In my opinion, it is better to have the CI action be run as non-failing check even if it finds issues (some issues may need to be fixed and others ignored as it may not be applicable). We should at regular intervals (say once in 45 days) review for new issues discovered by the CI action and determine if it needs a fix. This would also serve as a feedback/educations to developers if their PR introduced any vulnerabilities unknowingly.

**Comment by danwinship** [comment]
> we should have been doing:
> ioutil.ReadFile(filepath.Clean(pidfile))

**NO**. We shouldn't have. For any value of `pidfile`, `ioutil.ReadFile(pidfile)` and `ioutil.ReadFile(filepath.Clean(pidfile))` will read the same file.

What they _meant_ is that if you are combining together a base path and a user-provided path, you should sanitize _the user-provided part only_ before combining it with the base path, to ensure that the user-provided part doesn't start with `"../"` and subvert your base path.

But if you do

    pidfile := "/var/run/ovn-kubernetes/" + componentName + ".pid"
    bytes, err := ioutil.ReadFile(filepath.Clean(pidfile))

like they suggest, and `componentName` is `"../sshd.pid"`, then all it means is that you'll be calling `ioutil.ReadFile("/var/run/sshd.pid")` instead of `ioutil.ReadFile("/var/run/ovn-kubernetes/../sshd.pid")`, which means you're saving the kernel the work of having to resolve the `..` itself, but still falling for the attack.

This is why I hate security checkers.

**Comment by trozet** [comment]
@danwinship yeah that's why i was suggesting if it was a user-provided config file, but I'm not even sure if that is a concern.

I think for me at this point I don't see much value in including this, other than it could be another thing that breaks our CI. I think if @vjayaramrh you run this every so often and end up hitting a security bug that this tool actually caught, then I would be much more inclined at that point to enable it.

**Comment by vjayaramrh** [comment]
> @danwinship yeah that's why i was suggesting if it was a user-provided config file, but I'm not even sure if that is a concern.
> 
> I think for me at this point I don't see much value in including this, other than it could be another thing that breaks our CI. I think if @vjayaramrh you run this every so often and end up hitting a security bug that this tool actually caught, then I would be much more inclined at that point to enable it.

FYI: The action has been added not to break the CI even if it finds issues. In my opinion, having it integrated it as a CI action (action being set NOT TO FAIL when it finds issues) is the way to run it often. I am open to discussing alternate ways to run this tool or an alternate tool that gives feedback to developers that they are following secure coding practices.

**Comment by dave-tucker** [comment]
If it's running in CI and nobody's looking at the results there is no point in running it.
I'd suggest manually running it, opening issues for useful things it finds, and if we the trend is that we see more signal than noise (and can filter that noise) then I'd be open to adding it to CI.

**Comment by trozet** [comment]
We could never come to an agreement that we should include this as part of CI, so closing this for now.

**Pull request closed without merging**